src/types.rs

```diff
@@ -475,3 +506,209 @@ impl<'buf> RecvMsgOut<'buf> {
         self.header.flags
     }
 }
+
+/// [MatchOn] is a builder for constructing match criteria for request cancelation.
+///
+/// Requests can be canceled by matching on the user_data field and/or the file descriptor.
+///
+/// ```
+/// use io_uring::types;
+///
+/// // Match any request referencing file descriptor 42.
+/// types::MatchOn::fd(types::Fd(42)).match_any();
+/// ```
+#[derive(Debug)]
+pub struct MatchOn {
+    pub(crate) user_data: Option<u64>,
+    pub(crate) fd: Option<sealed::Target>,
+    multiple: MatchMultiple,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum MatchMultiple {
+    /// Match all requests. This indiscriminately cancels all requests,
+    /// regardless of user_data or fd.
+    All,
```

Comment:
You and I might have different understandings of the `all` match.

From the man page:
```
IORING_ASYNC_CANCEL_ALL
              Cancel all requests that match the given criteria, rather than just  can‐
              celing the first one found. Available since 5.19.
```

In fact, we might have different interpretations for what combinations are allowed. The liburing helper functions allow matching on user_data or fd, not both.

The `all` means all with the same user_data or all with the same fd.

Or the `any` flag means don't bother looking at the user_data, just match any, (and I'm assuming) if `any` and `all` are specified, it means all in-flight operations. It would be very odd to ask for `any` and just have the first matched, but that would seem to be allowed.

Reply:
```
CancelBuilder::new() // match on any
user_data(mut self, user_data) -> Self // match on user_data rather than any
fd(mut self, fd) -> Self // match on fd rather than any
all(mut self) -> Self // match on all rather than just the first.
```
So I think the builder impl could have just the four functions.

Reply:
Thanks, I think I was swapping around ALL and ANY here.

The following statements seem true based on my testing and read of the manpage.
1. Supplying a FD argument requires setting either **one of** sys::IORING_ASYNC_CANCEL_FD or sys::IORING_ASYNC_CANCEL_FD_FIXED.
2. Supplying a FD argument "overrides" the userdata argument. Testing shows that if you supply both, the FD argument takes precedence. 
3. `IORING_ASYNC_CANCEL_ALL` will cancel all requests which match either the FD argument or the user data argument.
4. `IORING_ASYNC_CANCEL_ANY` will cancel all operations, regardless of the other flags that were set. 

This would make the API you suggested output flags like the following:

- `CancelBuilder::new()` -> IORING_ASYNC_CANCEL_ANY: Cancel everything
- `CancelBuilder::user_data()` -> no flags: Cancel one request matching user_data.
- `CancelBuilder::fd()` -> IORING_ASYNC_CANCEL_FD: Cancel one request matching the FD or the fixed FD for FD_FIXED.
- `CancelBuilder::all()` previous flags | IORING_ASYNC_CANCEL_ALL: Cancel all matching requests. 


I'll implement this and try to beef up the test suite to check these combinations. Thanks again!

Reply:
yes, I concur with your descriptions. Bear in mind quininer hasn't sanctioned any of this as their timezone is completely different. But the Timespec has a small builder associated with it, and the semantics of the cancel are interesting enough that I would say the builder is worth it here too.

As for your point 1, I wouldn't be surprised if IORING_ASYNC_CANCEL_FD was supposed to be set for either type, regular and fixed, and only the *FD_FIXED flag should be added to make it the fixed version. But maybe the kernel accepts it both ways.